Add got-kills.md file (#115)
memgraph/docs: products/memgraph/upcoming/tutorials/got-deaths.md

## Game of Thrones deaths

This article is a part of a series intended to show how to use Memgraph
on real-world data to retrieve some interesting and useful
information.

We highly recommend checking out the other articles from this series which
are listed in our [tutorial overview section](tutorials-overview.md).

### Introduction

**WARNING** - If you've been living under a rock and still have not watched Game of Thrones,
this tutorial could contain ***spoilers***. 

We can all agree that 2020. was one of the worst years in our lives because of the worldwide
pandemic, 2019. still comes close with that disappointing, poorly written ending to one of the
best shows of all time, Game of Thrones. Seven-year build-up to that awful last season crushed

> the souls of millions of fans around the world. Nevertheless, the Game of Thrones world is full

Reviewer: This is a cool intro but it is too subjective to be part of a technical tutorial. Tutorials should be written in an objective manner.

> of interesting characters, locations, scenarios, and most importantly deaths. If you want to know

Reviewer: `[...] and most importantly deaths.` It seems like that was the most important thing in GoT, but it isn't. Maybe rewrite this in such a sense where you reference sudden deaths like Ned Stark and how it changed that feeling that main characters will be alive till the end so on so on...

how many characters would survive if Jon Snow stayed dead, which House has the best Kill/Death Ratio

> or who was the biggest traitor in the show, you have come to the right place!

Reviewer: ```suggestion
or who was the biggest traitor in the show, you came to the right place!

```

### Data Model

> Although you know the Game of Thrones TV show is based on series of books called A Song of Ice and

Reviewer: Try to skip `you know` when doing tech documentation except when referencing material in paragraph like `As you probably notes in the previous example`, `You could see that this code throws a new error`. Try to skip general knowledge "you know"s.

> Fire, our graph database contains only characters from said TV shows as books are still not finished.

Reviewer: `[...] from said TV shows [...]` - said is a strange word here.

This tutorial would not be possible with data analyst David Murphy who shared his collection of
on-screen deaths on [this link](https://data.world/datasaurusrex/game-of-thones-deaths). Some of the rules you have to take into concern are that kills that

> happened off-screen and undead(wraiths) are not counted. The dataset we used was slightly modified, to 

Reviewer: Try to simplify sentences. This is complex:
`Some of the rules you have to take into concern are that kills that happened off-screen and undead(wraiths) are not counted`

This is simpler, yet gives same message:
`We won't we working with kills and deaths that happened off-screen or were tied to undeads (wraiths).`

> be precise, columns "Episode Name" and "IMDb Rating" were added.

Reviewer: `to be precise` can be removed.

> Now, the data we'll be using in our queries can be classified as follows:

Reviewer: `Now` would make sense if this was step number X, but it isn't, so simply saying `Graph model consists of` makes things simple and readable.

>   * nodes, labeled as "Character", "Allegiance", "Death", "Episode", "Season", or "Location"

Reviewer: You already have a list of node labels below so this is insufficient. Check here how it is done (https://docs.memgraph.com/memgraph/tutorials-overview/football-transfers#data-model) from point of structure, formatting edge types and node labels as well as short description and example.

    * a "Character" node has a "name" attribute corresponding to the character's name (e.g. "Castle Black")
  * edges, of type "KilledBy", "Killed","LoyalTo", "VictimIn","KillerIn", "InEpisode", "InSeason", or "InLocation"
    * edges "KilledBy" and "Killed" connect two Character nodes and they have 2 attributes, 
      "method" which says how was the character killed (e.g. "Knife") and "count" attributes which says
      how many of these characters was killed (e.g. if "Jon Snow" killed 10 "Lannister soldiers" then on 
      this path "count" would be "10")
    * edge "LoyalTo" connects "Character" node with "Allegiance" node
    * edges "VictimIn" and "KillerIn" connects "Character" node with "Death" node
    * edge "InLocation" connects node "Death" with "Location" node
    * edge "InEpisode" connects node "Death" node "Episode"
    * edge "InSeason" connects nodes "Death" and "Episode" with "Season" node

### Importing the Snapshot

We have prepared a database snapshot for this example, so the user can easily
import it when starting Memgraph using the `--data-directory` option.

```plaintext
sudo -u memgraph \

>   /usr/lib/memgraph/memgraph --data-directory /usr/share/memgraph/examples/Got \

Reviewer: Because it is an acronym, it should be either `GOT` or `got`.

    --storage-snapshot-on-exit=false --storage-properties-on-edges=true
```

When using Memgraph installed from DEB or RPM package, the currently running
Memgraph server may need to be stopped before importing the example. The user
can do so using the following command:

```plaintext
systemctl stop memgraph
```

When using Docker, the example can be imported with the following command:

```plaintext
docker run -p 7687:7687 \
  -v mg_lib:/var/lib/memgraph -v mg_log:/var/log/memgraph -v mg_etc:/etc/memgraph \
  memgraph --data-directory /usr/share/memgraph/examples/Got \
  --storage-snapshot-interval-sec=0 --storage-wal-enabled=false \
  --storage-snapshot-on-exit=false --storage-properties-on-edges=true
```

The user should note that any modifications of the database state will persist
only during this run of Memgraph.

### Example Queries using OpenCypher

In the queries below, we are, as usual, using [OpenCypher](https://www.opencypher.org)
to query Memgraph via the console.

Here are some queries you might find interesting:

**MINI-GAME** - If you have watched the TV Show, try to guess each result before executing the query!

> 1) Let's start with a couple of simple queries. List the locations with the most deaths occurred.

Reviewer: ```suggestion
1) Let's start with a couple of simple queries. List the locations where most deaths occurred.

```

Can you guess which one is it?

```opencypher

> MATCH (l:Location)<-[:InLocation]-(d:Death) 

Reviewer: Edge types (even though it is not like that in old tutorials) should be in `SNAKE_UPPERCASE`

return l.name AS LocationName, COUNT(d) AS DeathCount
ORDER BY DeathCount DESC
```

2) Now that we have the location with the most deaths, let's list the episodes with the most deaths
as well.

```opencypher

> MATCH (d:Death)-[:InEpisode]->(e:Episode)

Reviewer: To make model simpler, it is good to skip word of the next label in edge type whenever possible. This `()-[]-()` should be read as a sentence where this one, and previous one sounds like `Death In Episode In Episode`. 

I would suggest `:HAPPENED_IN` and that one is used from Death to both Location and Episode. Because that way I can query `(d:Death { value: "something" })-[:HAPPENED_IN]-(n)` and in `n` I can get both Location and Episode. If I want only one of it I would add one change `(n:Location)`.

RETURN e.name, COUNT(d) AS KillCount
ORDER BY KillCount DESC
```

3) List the number of kills per season. I think we all know the highest count one. 

```opencypher

> MATCH (d:Death)-[:InSeason]->(s:Season) 

Reviewer: Same comment as before.

> RETURN s.number AS SeasonNumber, COUNT(d) as DeathNumber

Reviewer: On one occasion you have `XCount`, on another `XNumber` and I saw below just `X` when we are talking about `COUNT(n)`. Make things consistent, pick one and stick to that. `XCount` maybe makes more sense, e.g. `DeathCount`.

> RETURN s.number AS SeasonNumber, COUNT(d) as DeathNumber

Reviewer: Also sometimes you use `as` as `AS`, sometimes `as` as lowercased. Consistency. Pick any, but be consistent across this tutorial.

ORDER BY SeasonNumber ASC
```

> 4) While we know which season was by far the worst, can you guess the best one based on user ratings?

Reviewer: `While we know which season was by far the worst` - again doing we all know on general terms not present in the tutorial.

Let's list the seasons by IMDB ratings. The problem we get with using the `AVG()` function is that it
gives us too many decimals, therefore a useful solution is given in this example using `ROUND()`.

```opencypher

> MATCH (e:Episode)-[:InSeason]->(s:Season) 

Reviewer: Same comment as before for `InSeason`

RETURN s.number AS SeasonName, ROUND(100 * AVG(e.imdb_rating))/100 AS Rating
ORDER BY Rating DESC
```

> 5) There are many methods characters died with such as with sword or Dragonfire, but let's list victims

Reviewer: This first part of the sentence is complex and hard to read. Try to make it better and better sounding.

of unique methods.

```opencypher

> MATCH (c:Character)-[k:Killed]->(c2:Character)

Reviewer: You can remove variables `c` and `c2` as they are not used here.

WITH k.method AS Method, COUNT(k.method) AS MethodCount
WHERE MethodCount < 2
MATCH (c:Character)-[k:Killed]->(c2:Character)
WHERE k.method = Method
RETURN Method, c2.name as Victim
```

6) Daenerys Stormborn of House Targaryen, the First of Her Name, Queen of the Andals and the First Men,
Protector of the Seven Kingdoms, the Mother of Dragons, the Khaleesi of the Great Grass Sea, the Unburnt,
the Breaker of Chains or shortened to "Daenerys Targaryen" in our database is by far the biggest killer on
the show. Let's list all the episodes she killed in as well as characters she killed.

```opencypher
MATCH (c:Character {name: 'Daenerys Targaryen'})-[k:Killed]->(c2:Character)
MATCH (c)-[:KillerIn]->(d:Death)<-[:VictimIn]-(c2)
MATCH (d)-[:InEpisode]-(e:Episode)

> return distinct(c2.name) as Victim, COUNT(d) as Kills, e.name

Reviewer: return and distinct are lowercased here while all keywords in the tutorial and UPPERCASE. Make things consistent. `e.name` - You always did AS except now, why is that missing? Consistency again.

ORDER BY Kills DESC
```

7) Remember that shocking last episode of the fifth season when they killed Jon Snow and we totally thought
he was gonna stay dead? Well, let's list all the characters that would survive if he actually stayed dead.

```opencypher
MATCH (c:Character {name: 'Jon Snow'})-[k:Killed]->(c2:Character)
MATCH (c)-[:VictimIn]->(d2:Death)

> MATCH (c)-[:KillerIn]->(d:Death)<-[:VictimIn]-(c2)

Reviewer: WHen you have complex queries, make use of variable names, like `(jon:Character { name: 'Jon Snow')}-[:Killed]-(:killed_by_jon:Character)`

Also make sure to clean all Cyphers from variables that you don't use like `k` in `Killed`.

WHERE d.order>d2.order
return distinct(c2.name) as Victim, COUNT(d) as Kills
ORDER BY Kills DESC
```

8) Houses or allegiances are one of the main aspects of Westeros. We know some houses killed more characters
than others, but that doesn't matter in the end, what matters is efficiency. Let's list the allegiances with

> the best Kill/Death Ratios. Here we came across one more problem, if an allegiance had more deaths than kills,

Reviewer: ```suggestion
the best Kill/Death Ratios. Here we came across one additional problem, if an allegiance had more deaths than kills,

```

> the KD would be 0. This can easily be fixed with the `toFloat()` function.

Reviewer: When introducing new words like `KD` make sure to have intro to the reader about it. That means adding it to previous sentence `Kill/Death Rations, or short KD ratios`. Also you are missing `ratio`, `KD ratio would be 0`. As KD is not used so much except in one additional place, I would keep full working in second place.

```opencypher

> MATCH (c:Character)-[death:Killed]->(c2:Character)-[l:LoyalTo]->(a:Allegiance)

Reviewer: I won't adding similar comments below for consistency and cleanup. Those comments are up there and same applies to most of the queries.

MATCH (c:Character)-[:LoyalTo]->(a:Allegiance)
MATCH (c)-[:VictimIn]-(d:Death)-[:InEpisode]-(e:Episode {name: 'Battle of the Bastards'})
RETURN a.name AS HouseName, COUNT(d) AS DeathCount
ORDER BY DeathCount DESC
LIMIT 2;
```

10) Who do you think was the biggest traitor in terms of killing in its own allegiance? Well, let's check it out!

```opencypher
MATCH (c:Character)-[:Killed]->(c2:Character)
MATCH (c)-[:LoyalTo]->(a:Allegiance)<-[:LoyalTo]-(c2)
RETURN c.name AS Traitor, COUNT(c2) AS Kills
ORDER BY Kills DESC
```

11) One of the biggest features of Memgraph is drawing the graphs of queries we execute. Let's visualize all the
Loyalties with Characters. Execute the following query and head out to `GRAPH` tab.

```opencypher
match (c:Character)-[i:LoyalTo]-(l) return c,i,l;
```

12) Let's see how it looks like if we want to visualize Jon Snow kills with its locations.

```opencypher
MATCH (c:Character {name: 'Jon Snow'})-[k:Killed]->(c2:Character)
MATCH (c)-[:KillerIn]->(d:Death)<-[di:VictimIn]-(c2)
MATCH (d)-[i:InLocation]->(l:Location)
return c2,di,d,i,l
```

13) Memgraph supports graph algorithms as well. Let's use Dijkstra's shortest path algorithm to show the most

> gruesome path of kills.

Reviewer: Explain what a path of kills is. Otherwise, it is left for a reader to define it.

```opencypher
MATCH p = (:Character)-[:Killed * wShortest (e,v | e.count) total_weight]->(:Character)
RETURN nodes(p) as Kills, total_weight

> ORDER BY total_weight DESC

Reviewer: Now naming is in snake_case. Consistency again. Also make a better name rather than `total_weight`.

## Game of Thrones deaths

This article is a part of a series intended to show how to use Memgraph
on real-world data to retrieve some interesting and useful
information.

We highly recommend checking out the other articles from this series which
are listed in our [tutorial overview section](tutorials-overview.md).

### Introduction

**WARNING** - this tutorial could contain Game of Thrones ***spoilers***.  

Game of Thrones is an American fantasy drama television series created by David Benioff and D. B. Weiss for HBO.
It is an adaptation of A Song of Ice and Fire, George R. R. Martin's series of fantasy novels,  the first of which is
A Game of Thrones. The Game of Thrones world is full of interesting characters, locations, scenarios, unexpected deaths, and plot twists.

Even though, the COVID-19 pandemic hit the entire world in 2020 and is now starting to become one of the worst
years in recent history, 2019 was also a huge disappointment to all the Game of Thrones fans. According to IMDb ratings, a seven-year 

> build-up resulted in a poorly written ending of the last season and ruined the ending of one of the most popular shows on the planet.  

Reviewer 2: This is still a bit opinionated. IMDb ratings don't necessarily imply poor writing, but rather poor audience reception.
We aren't an authority on screenplay writing and we shouldn't be stating that something is poorly written. (Although I completely aggree with you :wink:)

Nonetheless, if you want to know how many characters would have survived if Jon Snow had stayed dead, which House had the best Kill/Death Ratio

> or who was the biggest traitor in the show, you came to the right place!

Reviewer 2: I see you've used the oxford comma before. That's great! You can use it here as well: `... Ratio, or who ...`

### Data Model

Although the Game of Thrones TV show is based on series of books, our graph database contains only characters from 

> previously mentioned TV shows as books are still not finished. This tutorial would not be possible with data analyst David 

Reviewer 2: `books` => `the books`?

> previously mentioned TV shows as the books are still not finished. This tutorial would not be possible with data analyst David 

Reviewer 2: _... only characters from **the** previously mentioned TV show ..._

Also

_This tutorial would not be possible **without** data analyst David_

> Murphy who shared his collection of on-screen deaths on [this link](https://data.world/datasaurusrex/game-of-thones-deaths).We won't we working with kills and deaths that happened off-screen or were 

Reviewer 2: Typo: _We won't **be** working_

Also, no space before `We`

> Murphy who shared his collection of on-screen deaths on [this link](https://data.world/datasaurusrex/game-of-thones-deaths). We won't be working with kills and deaths that happened off-screen or were 

Reviewer 2: `this link` -> Did he also maybe have a blog post? You can expand on this a bit instead of simply adding "this link".
It would be cool if we shared his study, blog post, or website. Maybe adding a new sentence such as `.. for more information you can visit xxx ...`, or maybe give a short intro before providing the link.

Also, I know you mentioned that we need his permission. Did we contact him? @tonilastre Opinions?

Author: We don't actually need his permission, he just said `"If you create something from this data, please get in touch and share your creations with me on social media. I’d love to see and showcase them!"`. It would be great if we contacted him! He has a good following as well.

> tied to undeads (wraiths). The dataset we used was slightly modified, columns "Episode Name" and "IMDb Rating" were added.

Reviewer 2: Please check if **undeads** is a word. I usually hear **the undead** if they are in plural (were tied to the undead). If it's okay, then just ignore this comment. :smile:

Author: https://en.wiktionary.org/wiki/undead#:~:text=undead%20(plural%20undeads) according to the internet, both are correct.

> tied to the undead (wraiths). The dataset we used was slightly modified, columns "Episode Name" and "IMDb Rating" were added.

Reviewer 2: _(optional) opinion:_ This sentence is a bit clumsy. I think it would sound better if you replaced the comma with some sort of conjunction or split the sentences with a full-stop.

The model consists of the following nodes:
* a `Character` node has a "name" attribute corresponding to the character's name (e.g. `"Jon Snow"`)
* an `Allegiance` node has a "name" attribute corresponding to the house name the character 
    is loyal to (e.g.  `"House Stark"`)
* a `Death` node has an "order" attribute corresponding to the order in which the death happened 

>     in show (e.g. `602`)

Reviewer 2: `show` => `the show`?

> * an `Episode` node has a "number" attribute corresponding to the number of episodes (e.g. `10`), 

Reviewer 2: Might as well use \`number\` instead of "number" so the markdown can render code.

  memgraph --data-directory /usr/share/memgraph/examples/got \
  --storage-snapshot-interval-sec=0 --storage-wal-enabled=false \
  --storage-snapshot-on-exit=false --storage-properties-on-edges=true
```

The user should note that any modifications of the database state will persist
only during this run of Memgraph.

### Example Queries using OpenCypher

In the queries below, we are, as usual, using [OpenCypher](https://www.opencypher.org)
to query Memgraph via the console.

Here are some queries you might find interesting:

**MINI-GAME** - If you have watched the TV Show, try to guess each result before executing the query!
1) Let's start with a couple of simple queries. List the locations where most deaths occurred.
Can you guess which one is it?

```opencypher
MATCH (l:Location)<-[:HAPPENED_IN]-(d:Death) 
RETURN l.name AS location_name, COUNT(d) AS death_count
ORDER BY death_count DESC
```

2) Now that we have the location with the most deaths, let's list the episodes with the most deaths
as well.

```opencypher
MATCH (d:Death)-[:HAPPENED_IN]->(e:Episode)
RETURN e.name AS episode_name, COUNT(d) AS kill_count
ORDER BY kill_count DESC
```

> 3) List the number of kills per season. If you have watched the show, you should guess this one.

Reviewer 2: you **should be able** to guess this one

```opencypher
MATCH (d:Death)-[:HAPPENED_IN]->(s:Season) 
RETURN s.number AS season_number, COUNT(d) AS death_count
ORDER BY season_number ASC
```

> 4) Worst season by far, based on user ratings, is the last one, but can you guess the best one?

Reviewer 2: _Worst_ => _The worst ..._
Also, here we have that opinionation again. Maybe consider using something along the lines of: `The most poorly rated season by far was the last one, ...`

Let's list the seasons by IMDB ratings. The problem we get with using the `AVG()` function is that it
gives us too many decimals, therefore a useful solution is given in this example using `ROUND()`.

```opencypher
MATCH (e:Episode)-[:PART_OF]->(s:Season) 
RETURN s.number AS season_name, ROUND(100 * AVG(e.imdb_rating))/100 AS rating
ORDER BY rating DESC
```

5) There are many methods by which characters were killed such as sword or Dragonfire, but let's list victims
of unique methods.

```opencypher
MATCH (:Character)-[k:KILLED]->(:Character)
WITH k.method AS kill_method, COUNT(k.method) AS method_count
WHERE method_count < 2
MATCH (killer:Character)-[k:KILLED]->(victim:Character)
WHERE k.method = kill_method
RETURN kill_method, victim.name AS victim
```

6) Daenerys Stormborn of House Targaryen, the First of Her Name, Queen of the Andals and the First Men,
Protector of the Seven Kingdoms, the Mother of Dragons, the Khaleesi of the Great Grass Sea, the Unburnt,

> the Breaker of Chains or shortened to "Daenerys Targaryen" in our database is the biggest killer on

Reviewer 2: :rofl: :+1:

the show. Let's list all the episodes she killed in as well as characters she killed.

```opencypher
MATCH (daenerys:Character {name: 'Daenerys Targaryen'})-[:KILLED]->(victim:Character)
MATCH (daenerys)-[:KILLED_IN]->(d:Death)<-[:DIED_IN]-(victim)
MATCH (d)-[:HAPPENED_IN]-(e:Episode)
RETURN DISTINCT(victim.name) AS victim, COUNT(d) AS kill_count, e.name AS episode_name
ORDER BY kill_count DESC
```

7) Remember that shocking last episode of the fifth season when they killed Jon Snow and we totally thought
he was gonna stay dead? Well, let's list all the characters that would survive if he actually stayed dead.

```opencypher
MATCH (jon:Character {name: 'Jon Snow'})-[:KILLED]->(victim:Character)
MATCH (jon)-[:DIED_IN]->(jon_death:Death)
MATCH (jon)-[:KILLED_IN]->(victim_death:Death)<-[:DIED_IN]-(victim)
WHERE victim_death.order>jon_death.order
RETURN DISTINCT(victim.name) AS victim, COUNT(victim_death) AS kill_count
ORDER BY kill_count DESC
```

8) Houses or allegiances are one of the main aspects of Westeros. Some houses killed more characters
than others, but that doesn't matter in the end, what matters is efficiency. Let's list the allegiances with

> the best Kill/Death Ratios, or short KDR. Here we came across one additional problem, if an allegiance had more 

Reviewer 2: `or short KDR` => `or KDR for short`

Also in `... problem, if an ...`, you can replace the comma with a full-stop and start a new sentence.

RETURN a.name AS allegiance_name, SUM(kill.count) AS kills, deaths, ROUND(100 *(TOFLOAT(SUM(kill.count))/deaths))/100 AS KDR
ORDER BY KDR DESC;
```

9) One of the best-rated episodes, Battle of the Bastards, showed us a fight between two houses: Stark and Bolton.
Let's see which one had more casualties.

```opencypher
MATCH (c:Character)-[:LOYAL_TO]->(a:Allegiance)
MATCH (c)-[:DIED_IN]-(d:Death)-[:HAPPENED_IN]-(:Episode {name: 'Battle of the Bastards'})
RETURN a.name AS house_name, COUNT(d) AS death_count
ORDER BY death_count DESC
LIMIT 2;
```

10) Who do you think was the biggest traitor in terms of killing in its own allegiance? Well, let's check it out!

```opencypher
MATCH (killer:Character)-[:KILLED]->(victim:Character)
MATCH (killer)-[:LOYAL_TO]->(a:Allegiance)<-[:LOYAL_TO]-(victim)
RETURN killer.name AS traitor, COUNT(victim) AS kill_count
ORDER BY kill_count DESC
```

11) One of the biggest features of Memgraph is drawing the graphs of queries we execute. Let's visualize all the
Loyalties with Characters. Execute the following query and head out to `GRAPH` tab.

```opencypher
MATCH (c:Character)-[i:LOYAL_TO]-(l) RETURN c,i,l;
```

> 12) Let's see how it looks like if we want to visualize Jon Snow kills with its locations.

Reviewer 2: Jon Snow's kills with **their** locations.

```opencypher
MATCH (jon:Character {name: 'Jon Snow'})-[:KILLED]->(victim:Character)
MATCH (jon)-[:KILLED_IN]->(death:Death)<-[victim_to_death:DIED_IN]-(victim)
MATCH (death)-[death_to_location:HAPPENED_IN]->(location:Location)
RETURN victim,victim_to_death,death,death_to_location,location
```

13) Memgraph supports graph algorithms as well. Let's use Dijkstra's shortest path algorithm to show the most

> gruesome path of kills. Example of path of kills is: `Jon Snow` killed `5` `Lannister Soldiers` and they killed

Reviewer 2: `Example` => `An example kill path ...`

on the show deaths. We won't be working with kills and deaths that happened off-screen or were tied to the undead (wraiths).
The dataset we used was slightly modified, in which columns "Episode Name" and "IMDb Rating" were added.

The model consists of the following nodes:
* a `Character` node has a `name` attribute corresponding to the character's name (e.g. `"Jon Snow"`)
* an `Allegiance` node has a `name` attribute corresponding to the house name the character 
    is loyal to (e.g.  `"House Stark"`)
* a `Death` node has an `order` attribute corresponding to the order in which the death happened 
    in the show (e.g. `602`)
* an `Episode` node has a `number` attribute corresponding to the number of episodes (e.g. `10`), 
    `name` attribute corresponding to the name of the episode (e.g. `"Mothers Mercy"`) and `imdb_rating`
    episode corresponding to the IMDB rating of the episode (e.g. "9.1")
* a `Season` node has a `number` attribute corresponding to the number of the season (e.g. `10`)
* a `Character` node has a `name` attribute corresponding to the character's name (e.g. `"Castle Black"`)

Nodes are connected with the following edges:

> * `:KILLED_BY` and `:KILLED` - connect two Character nodes and they have 2 attributes, 

Reviewer: Btw why do we need both edge types?

  `method` which says how was the character killed (e.g. `"Knife"`) and `count` attribute representing a
  number of how many of these characters were killed (e.g. if `"Jon Snow"` killed `10` `"Lannister soldiers"` 

>   then on this path `count` would be `10`)

Reviewer: ```suggestion
  then on this edge `count` would be `10`)

```

* `:LOYAL_TO` - connects `Character` node with `Allegiance` node representing an allegiance the character is
  loyal to
* `:VICTIM_IN` and `:KILLER_IN` - connects `Character` node with `Death` node in which death happened 
* `:HAPPENED_IN` - connects node `Death` with `Episode`, `Season` and `Location` nodes representing details

>   of death

Reviewer: ```suggestion
  of the death

```

MATCH (:Character)<-[kill:KILLED]-(:Character)-[:LOYAL_TO]->(a)
RETURN a.name AS allegiance_name, SUM(kill.count) AS kills, deaths, ROUND(100 *(TOFLOAT(SUM(kill.count))/deaths))/100 AS KDR
ORDER BY KDR DESC;
```

9) One of the best-rated episodes, Battle of the Bastards, showed us a fight between two houses: Stark and Bolton.
Let's see which one had more casualties.

```opencypher
MATCH (c:Character)-[:LOYAL_TO]->(a:Allegiance)
MATCH (c)-[:VICTIM_IN]-(d:Death)-[:HAPPENED_IN]-(:Episode {name: 'Battle of the Bastards'})
RETURN a.name AS house_name, COUNT(d) AS death_count
ORDER BY death_count DESC
LIMIT 2;
```

10) Who do you think was the biggest traitor in terms of killing in its own allegiance? Well, let's check it out!

```opencypher
MATCH (killer:Character)-[:KILLED]->(victim:Character)
MATCH (killer)-[:LOYAL_TO]->(a:Allegiance)<-[:LOYAL_TO]-(victim)
RETURN killer.name AS traitor, COUNT(victim) AS kill_count
ORDER BY kill_count DESC
```

11) One of the biggest features of Memgraph is drawing the graphs of queries we execute. Let's visualize all the
Loyalties with Characters. Execute the following query and head out to the `GRAPH` tab.

```opencypher

> MATCH (c:Character)-[i:LOYAL_TO]-(l) RETURN c,i,l;

Reviewer: You could benefit from having a path variable here when you want to visualize. Check other tutorials for help (e.g. football transfers).

```

12) Let's see how it looks like if we want to visualize Jon Snow kills with their locations.

```opencypher
MATCH (jon:Character {name: 'Jon Snow'})-[:KILLED]->(victim:Character)
MATCH (jon)-[:KILLER_IN]->(death:Death)<-[victim_to_death:VICTIM_IN]-(victim)
MATCH (death)-[death_to_location:HAPPENED_IN]->(location:Location)

> RETURN victim,victim_to_death,death,death_to_location,location

Reviewer: Same as before, you could benefit from having a path variable here when you want to visualize. Check other tutorials for help (e.g. football transfers).

Reviewer: Or what David has suggested to you which will cover not just graph output, but also data view.
